clippy_lints/src/lifetimes_bound_nested_ref.rs

```diff
@@ -0,0 +1,222 @@
+/// Lints to help dealing with unsoundness due to a compiler bug described here:
```

Comment:
Starting a thread to localize this discussion:

My question is about the motivation of including this into Clippy: If Clippy is able to detect those unsoundness issues, shouldn't this be implemented in the compiler and fix the unsoundness bugs?

One explanation for this approach that I see is, that this lint will have false positives (FPs), so implementing it in the compiler would restrict the language too much. Am I right in that assumption?

If this should be the case, putting the lint into a warn-by-default group might be hard. We definitely should run our `lintcheck` tool to look for FPs in popular crates.

Reply:
> Starting a thread to localize this discussion:
> 
> My question is about the motivation of including this into Clippy: If Clippy is able to detect those unsoundness issues, shouldn't this be implemented in the compiler and fix the unsoundness bugs?

Ideally yes, and I hope the compiler gets fixed before this is ready. 
But the 25860 issue was reported in 2015...

> 
> One explanation for this approach that I see is, that this lint will have false positives (FPs), so implementing it in the compiler would restrict the language too much. Am I right in that assumption?

An implied bound is correct because it is implied by the compiler input code. Adding an implied bound is also correct, but it adds redundancy in the input code. I think there is no language restriction here.

Reply:
Yea, this isn't implemented in rustc, because it would require bounds in a lot of places, while the new solver that can fix this (citation needed) is still about a year out.

I just wonder how many ppl will use this lint. I don't want you to waste time writing a complex lint that then becomes mostly unused